extensions/wikia/Chat2/ChatBanListSpecial_helper.php

```diff
@@ -0,0 +1,269 @@
+<?php
+
+/**
+ * @package MediaWiki
+ * @subpackage SpecialPage
+ * @author Piotr Molski <moli@wikia.com>
+ * @version: $Id$
+ */
+
+use Wikia\Service\User\Permissions\PermissionsServiceAccessor;
+
+class ChatBanData
+{
+	use PermissionsServiceAccessor;
+
+	/**
+	 * @var string
+	 */
+	private $cityId;
+
+	/**
+	 * @var string
+	 */
+	private $userName;
+
+	/**
+	 * @var int
+	 */
+	private $limit;
+
+	/**
+	 * @var int
+	 */
+	private $offset;
+
+	/**
+	 * @var string
+	 */
+	private $order;
+
+	/**
+	 * @var array
+	 */
+	private $orderOptions;
+
+	/**
+	 * @var Object
+	 */
+	private $skin;
+
+	/**
+	 * Database
+	 * @var string
+	 */
+	private $db;
+
+	/**
+	 * @var string
+	 */
+	private $table;
+
+
+	function __construct( $city_id, $load = 1 ) {
+		global $wgExternalDatawareDB;
+
+		$this->cityId = $city_id;
+		$this->db = $wgExternalDatawareDB;
+		$this->table = 'chat_ban_users';
+		$this->skin = RequestContext::getMain()->getskin();
+
+		$this->orderOptions = [
+			'timestamp' => 'start_date %s',
+			'target'    => 'cbu_user_id %s',
+			'expires'   => 'end_date %s',
+			'blockedBy' => 'cbu_admin_user_id %s',
+		];
+
+		if ( $load == 1 ) {
+			$this->load();
+		}
+	}
+
+	function load() {
+		$this->setLimit();
+		$this->setOffset();
+		$this->setOrder();
+	}
+
+	function setUserName( $username = '' ) { $this->userName = $username; }
+
+	function setLimit( $limit = Listusers::DEF_LIMIT ) { $this->limit = $limit; }
+
+	function setOffset( $offset = 0 ) { $this->offset = $offset; }
+
+	function setOrder( $order = '' ) {
+		if ( empty( $order ) ) {
+			// default order
+			$order = 'timestamp:desc';
+		}
+
+		$validSortDirections = [
+			'asc',
+			'desc',
+		];
+
+		list ( $orderName, $orderDesc ) = explode( ":", $order );
+		if ( isset( $this->orderOptions[ $orderName ] ) && in_array( $orderDesc, $validSortDirections ) ) {
+			$this->order = sprintf( $this->orderOptions[ $orderName ], $orderDesc );
+		}
+	}
+
+	function getUserName() { return $this->userName; }
+
+	function getLimit() { return $this->limit; }
+
+	function getOffset() { return $this->offset; }
+
+	function getOrder() { return $this->order; }
+
+	public function loadData() {
+		global $wgMemc, $wgLang;
+		wfProfileIn( __METHOD__ );
+
+		/* initial values for result */
+		$data = [
+			'cnt'      => 0,
+			'sColumns' => implode( ",", array_keys( $this->orderOptions ) ),
+			'data'     => [ ],
+		];
+
+		$subMemkey = [
+			'O' . $this->offset,
+			'O' . $this->order,
+			'L' . $this->limit,
+			'U' . $this->userName,
+		];
+
+		$memkey = wfForeignMemcKey( $this->cityId, null, "cbdata", md5( implode( ', ', $subMemkey ) ) );
+		$cached = $wgMemc->get( $memkey );
```

Comment:
Instead of using `$wgMemc` we should consider caching this response in the controller instead and purging it if a ban is updated.